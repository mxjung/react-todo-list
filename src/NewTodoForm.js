import React, { useState } from "react";

/** Creates NewTodoForm component:
 *  This component should render a form with one text input for the task to be created. 
 *  When this form is submitted, a new Todo component should be created.
 * */

function NewTodoForm({addTodo}) {
  const INITIAL_STATE = { text: "" };
  const [formData, setFormData] = useState({...INITIAL_STATE});

  // IF: we hasn't spread {...INITIAL_STATE} into useState, and we had executed the line below, 
  // our INITIAL_STATE would be changed as well, since formData (if initial_state is not spread/slice)
  // would be pointing to the same place in memory as INITIAL STATE. Same reason why I spread Initial state
  // in handleSubmit();
  // IF BY ACCIDENT: formData.text = "hello there";
  
  /** Send {text} to parent
   *    & clear form. */

  const handleSubmit = evt => {
    evt.preventDefault();
    addTodo(formData);
    setFormData({...INITIAL_STATE});
  };

  /** Update local state w/curr state of input elem */

  const handleChange = evt => {
    const { name, value } = evt.target;
    setFormData(fData => ({
      ...fData,
      [name]: value
    }));
  };

  /** render form */

  // text isn't the best description ***** taskTitle, taskName, etc
  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="text">Todo:</label>
      <input
        id="text"
        name="text"
        value={formData.text}
        onChange={handleChange}
      />

      <button>Add a new Todo!</button>
    </form>
  );
}

export default NewTodoForm;