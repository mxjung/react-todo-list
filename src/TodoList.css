.todoApp {
  width: 600px;
  display: flex;
  flex-direction: column;
  background-color: azure;
  align-items: center;

}

.todolist {
  height: 400px;
  width: 400px;
  background-color: beige;
  position: relative;
  display: flex;
  flex-direction: column;
  margin: 30px;
}
