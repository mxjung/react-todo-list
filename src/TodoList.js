import React, { useState } from "react";
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';
import {v4 as uuid} from "uuid";
import './TodoList.css';

/**
 *  TodoList - this component should render the NewTodoForm component and should 
 *  render the list of Todo components. Place your state that contains all of the 
 *  todos in this component.
 */
function TodoList() {
  const [todos, setTodos] = useState([]);

  // Function that will be passed into NewTodoForm to update the todos state
  const addTodo = (todo) => {
    let newTodo = {
      ...todo,
      id: uuid(),
      text: todo.text
    };
    setTodos(todos => [...todos, newTodo]);
  }

  // Function that will be passed into Todo to delete Todo and update the todos state
  const removeTodo = (deleteId) => {
    setTodos(todos => {
      let oldTodos = [...todos];
      const newTodos = oldTodos.filter(targetTodo => 
        deleteId !== targetTodo.id
      );
      return newTodos;
    })
  }

  // Function that will be update List showing all todos
  const renderBoxes = () => {
    return (
      <div className="todolist">
      {todos.map(todo => (
        // <div key={todo.id}> // this would have worked, but no reason to, you can have it on Todo component
        // key has to go on what i map through
          <Todo key={todo.id} id={todo.id} text={todo.text} removeTodo={removeTodo}/>
        // </div>
      ))}
      </div>
    )
  }

  return (
    <div className="todoApp">
      <h2>Todo App</h2>
      {<NewTodoForm addTodo={addTodo} />}
      {renderBoxes()}
    </div>
  )
}

export default TodoList;