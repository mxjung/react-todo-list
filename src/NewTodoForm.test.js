import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";
import NewTodoForm from "./NewTodoForm";

it("renders without crashing", function () {
  render(<NewTodoForm />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<NewTodoForm />);
  expect(asFragment()).toMatchSnapshot();
})

it("Shows Form", function () {
  const { queryByLabelText } = render(<TodoList />);

  // Check for form
  const textInput = queryByLabelText("Todo:");
  expect(textInput).not.toBe(null);
})
