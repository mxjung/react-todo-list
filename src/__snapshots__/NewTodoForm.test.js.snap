// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="text"
    >
      Todo:
    </label>
    <input
      id="text"
      name="text"
      value=""
    />
    <button>
      Add a new Todo!
    </button>
  </form>
</DocumentFragment>
`;
